File: libs/utils/lib-app/src/main/java/com/yyxnb/what/app/AppUtils.java
package com.yyxnb.what.app;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Application;
import android.content.pm.ApplicationInfo;
import android.os.Bundle;
import android.util.Log;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;

/**
 * ================================================
 * 作    者：yyx
 * 日    期：2021/01/23
 * 描    述：AppUtils
 * ================================================
 */
public final class AppUtils {

    @SuppressLint("StaticFieldLeak")
    private static Application sApp;
    private static boolean sFirstRun = true;

    public static void init(final Application app) {
        if (app == null) {
            Log.e("AppUtils", "app is null.");
            return;
        }
        sFirstRun = false;
        if (sApp == null) {
            sApp = app;
            return;
        }
        if (sApp.equals(app)) {
            return;
        }
        sApp = app;
    }

    public static Application getApp() {

        if (sApp != null) {
            return sApp;
        }
        sApp = AppGlobals.getApplication();
        if (sApp == null) {
            throw new NullPointerException("application反射获取失败");
        }
        return sApp;
    }


    public static boolean isActivityAlive(final Activity activity) {
        return activity != null && !activity.isFinishing() && !activity.isDestroyed();
    }

    /**
     * 判断当前应用是否是debug状态
     */
    public static boolean isDebug() {
        try {
            ApplicationInfo info = getApp().getApplicationInfo();
            return (info.flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * 返回实例的泛型类型
     */
    @SuppressWarnings({"unchecked", "ConstantConditions"})
    public static <T> T getInstance(Object t, int i) {
        if (t != null) {
            try {
                return (T) ((ParameterizedType) t.getClass().getGenericSuperclass())
                        .getActualTypeArguments()[i];

            } catch (ClassCastException e) {
                e.printStackTrace();
            }
        }
        return null;
    }

    @SuppressWarnings({"unchecked", "ConstantConditions"})
    public static <T> Class<T> getClass(Object t) {
        // 通过反射 获取当前类的父类的泛型 (T) 对应 Class类
        return (Class<T>) ((ParameterizedType) t.getClass().getGenericSuperclass())
                .getActualTypeArguments()[0];
    }

    @SuppressWarnings("unchecked")
    public static <T> Class<T> getFiledClazz(Field field) {
        return (Class<T>) field.getGenericType();
    }

    public static boolean isFirstRun() {
        return sFirstRun;
    }

    public static class ActivityLifecycleCallbacks {

        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {/**/}

        public void onActivityStarted(Activity activity) {/**/}

        public void onActivityResumed(Activity activity) {/**/}

        public void onActivityPaused(Activity activity) {/**/}

        public void onActivityStopped(Activity activity) {/**/}

        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {/**/}

        public void onActivityDestroyed(Activity activity) {/**/}

    }
}

File: widget/src/main/java/com/yyxnb/widget/AppUtils.kt
package com.yyxnb.widget

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.pm.ApplicationInfo
import android.os.Bundle
import android.util.Log
import com.yyxnb.widget.AppGlobals.application
import java.lang.reflect.Field
import java.lang.reflect.ParameterizedType

/**
 * @author yyx
 */
object AppUtils {
    @SuppressLint("StaticFieldLeak")
    private var sApp: Application? = null

    fun init(app: Application?) {
        if (app == null) {
            Log.e("AppUtils", "app is null.")
            return
        }
        if (sApp == null) {
            sApp = app
            WidgetManager.init(sApp)
            return
        }
        if (sApp == app) {
            return
        }
        WidgetManager.unInit(sApp)
        sApp = app
        WidgetManager.init(sApp)
    }

    val app: Application
        get() {
            if (sApp != null) {
                return sApp!!
            }
            sApp = application
            if (sApp == null) {
                throw NullPointerException("application反射获取失败")
            }
            return sApp!!
        }

    fun isActivityAlive(activity: Activity): Boolean {
        return !activity.isFinishing && !activity.isDestroyed
    }

    /**
     * 判断当前应用是否是debug状态
     */
    val isDebug: Boolean
        get() = try {
            val info = app.applicationInfo
            info.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0
        } catch (e: Exception) {
            false
        }


    /**
     * 返回实例的泛型类型
     */
    fun <T> getInstance(t: Any?, i: Int): T? {
        if (t != null) {
            try {
                return (t.javaClass.genericSuperclass as ParameterizedType)
                        .actualTypeArguments[i] as T
            } catch (e: ClassCastException) {
                e.printStackTrace()
            }
        }
        return null
    }

    fun <T> getClass(t: Any): Class<T> {
        // 通过反射 获取当前类的父类的泛型 (T) 对应 Class类
        return (t.javaClass.genericSuperclass as ParameterizedType)
                .actualTypeArguments[0] as Class<T>
    }

    fun <T> getFiledClazz(field: Field): Class<T> {
        return field.genericType as Class<T>
    }


    open class ActivityLifecycleCallbacks {
        open fun onActivityCreated(activity: Activity?, savedInstanceState: Bundle?) { /**/
        }

        open fun onActivityStarted(activity: Activity?) { /**/
        }

        open fun onActivityResumed(activity: Activity?) { /**/
        }

        open fun onActivityPaused(activity: Activity?) { /**/
        }

        open fun onActivityStopped(activity: Activity?) { /**/
        }

        open fun onActivitySaveInstanceState(activity: Activity?, outState: Bundle?) { /**/
        }

        open fun onActivityDestroyed(activity: Activity?) { /**/
        }
    }
}